examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,356 @@
+;;;
+;;;
+;;; A Brainfold/Brainf*** compiler implemented in Coalton
+;;;
+;;; The compiler handles all standard bf commands:
+;;; ( > < + - . , [ ] ).
+;;;
+;;;
+;;; Run Brainfold programs with (run-program "++++[-]")
+;;;
+;;; or
+;;;
+;;; (run-file "/path/to/your/file.bf")
+;;;
+;;; try (coalton (hello-world)) in the REPL!
+;;;
+
+
+(cl:defpackage #:brainfold
+  (:use #:coalton
+        #:coalton-prelude)
+  (:local-nicknames (#:vec #:coalton-library/vector)
+		    (#:iter #:coalton-library/iterator)
+		    (#:cell #:coalton-library/cell)
+		    (#:char #:coalton-library/char)
+		    (#:str #:coalton-library/string)
+		    (#:list #:coalton-library/list)
+		    (#:arith #:coalton-library/math/arith))
+
+  (:export #:eval
+	   #:run-program
+	   #:run-file
+	   #:hello-world))
+
+;;;
+;;; Environment (Vector, Pointer, Print Buffer)
+;;;
+
+(in-package #:brainfold)
+
+(named-readtables:in-readtable coalton:coalton)
+
+(coalton-toplevel
+
+ ;;
+ ;; Initialize a vector with 0's, ideally 30k according to the
+ ;; specifications
+ ;;
+  
+ (declare bf-arr-size (UFix))
+ (define bf-arr-size 1000)
+
+ (define (generate-bf-array)
+   "Generates an empty vector to be inhabited by brainfold"
+   (let v = (vec:with-capacity bf-arr-size))
+   (iter:for-each! (fn (x)
```

Comment:
This can be written more concisely as `(vec:extend! v (iter:repeat-for 0 bf-array-size))`